cluster/src/shard_lock_manager.rs

```diff
@@ -0,0 +1,505 @@
+    define_result,
+    runtime::{JoinHandle, RuntimeRef},
+};
+use etcd_client::{
+    Client, Compare, CompareOp, LeaseKeepAliveStream, LeaseKeeper, PutOptions, Txn, TxnOp,
+};
+use log::{debug, error, info, warn};
+use prost::Message;
+use snafu::{ensure, Backtrace, ResultExt, Snafu};
+use tokio::sync::{oneshot, RwLock};
+
+#[derive(Debug, Snafu)]
+#[snafu(visibility = "pub")]
+pub enum Error {
+    #[snafu(display("Failed to keep alive, err:{source}.\nBacktrace:\n{backtrace:?}"))]
+    KeepAlive {
+        source: etcd_client::Error,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display(
+        "Failed to keep alive because of no resp, lease_id:{lease_id}.\nBacktrace:\n{backtrace:?}"
+    ))]
+    KeepAliveWithoutResp { lease_id: i64, backtrace: Backtrace },
+
+    #[snafu(display(
+        "Failed to grant lease, shard_id:{shard_id}, err:{source}.\nBacktrace:\n{backtrace:?}"
+    ))]
+    GrantLease {
+        shard_id: ShardId,
+        source: etcd_client::Error,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display("Failed to create lock in etcd, shard_id:{shard_id}, err:{source}.\nBacktrace:\n{backtrace:?}"))]
+    CreateLockInEtcd {
+        shard_id: ShardId,
+        source: etcd_client::Error,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display(
+        "Failed to execute txn of create lock, shard_id:{shard_id}.\nBacktrace:\n{backtrace:?}"
+    ))]
+    CreateLockTxn {
+        shard_id: ShardId,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display("Failed to revoke the lease, lease_id:{lease_id}, shard_id:{shard_id}, err:{source}.\nBacktrace:\n{backtrace:?}"))]
+    RevokeLease {
+        lease_id: i64,
+        shard_id: ShardId,
+        source: etcd_client::Error,
+        backtrace: Backtrace,
+    },
+}
+
+define_result!(Error);
+
+const DEFAULT_LEASE_SECS: u64 = 30;
+const LOCK_EXPIRE_CHECK_INTERVAL: Duration = Duration::from_millis(200);
+
+pub type ShardLockManagerRef = Arc<ShardLockManager>;
+
+/// Shard lock manager is implemented based on etcd.
+///
+/// Only with the lock held, the shard can be operated by this node.
+pub struct ShardLockManager {
+    key_prefix: String,
+    value: Bytes,
+    lease_ttl_sec: u64,
+
+    etcd_client: Client,
+    runtime: RuntimeRef,
+
+    // ShardID -> ShardLock
+    shard_locks: Arc<RwLock<HashMap<u32, ShardLock>>>,
+}
+
+/// The lease of the shard lock.
+struct Lease {
+    /// The lease id
+    id: i64,
+    /// The time to live of the lease
+    ttl: Duration,
+    /// Expired time in milliseconds
+    expired_at: Arc<AtomicU64>,
+
+    keep_alive_handle: Option<JoinHandle<()>>,
+}
+
+impl Lease {
+    fn new(id: i64, ttl: Duration) -> Self {
+        Self {
+            id,
+            ttl,
+            expired_at: Arc::new(AtomicU64::new(0)),
+            keep_alive_handle: None,
+        }
+    }
+
+    /// Check whether lease is expired.
+    fn is_expired(&self) -> bool {
+        let now = common_util::time::current_time_millis();
```

Comment:
This is not a monotonically clock, you should use Instant instead.
- https://docs.rs/chrono/latest/chrono/#date-and-time

Reply:
I guess the interval of expire check is not very accurate, so the time offsetting is allowed.